src/esssans/uncertainty.py

```diff
@@ -4,43 +4,63 @@
 
 from typing import Dict, TypeVar, Union, overload
 
+import numpy as np
 import scipp as sc
 
 T = TypeVar("T", bound=Union[sc.Variable, sc.DataArray])
 
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.Variable, sizes: Dict[str, int]
+    data: sc.Variable, template: sc.Variable
 ) -> sc.Variable:
     pass
 
 
 @overload
 def broadcast_with_upper_bound_variances(
-    data: sc.DataArray, sizes: Dict[str, int]
+    data: sc.DataArray, template: sc.DataArray
 ) -> sc.DataArray:
     pass
 
 
 def broadcast_with_upper_bound_variances(
-    data: Union[sc.Variable, sc.DataArray], sizes: Dict[str, int]
+    data: Union[sc.Variable, sc.DataArray], template: Union[sc.Variable, sc.DataArray]
 ) -> Union[sc.Variable, sc.DataArray]:
-    if _no_variance_broadcast(data, sizes):
+    """
+    Upper-bound estimate for errors from broadcasting a dense array to match the sizes
+    of a given template array.
+
+    We first count the number of non-masked elements in the template. To get the
+    variances multiplier, we then divide that number by the volume of the dimensions of
+    the input data array that are also found in the dimensions of the template. That
+    scaling factor is used to multiply the variances of the input data array.
+    Finally, an explicit broadcast is performed to bypass Scipp's safety check on
+    broadcasting variances.
+
+    Parameters
+    ----------
+    data
+        The data array to be normalized.
+    template
+        The template data array.
+    """
+    if _no_variance_broadcast(data, template.sizes):
         return data
-    size = 1
-    for dim, dim_size in sizes.items():
-        if dim not in data.dims:
-            size *= dim_size
     data = data.copy()
-    data.variances *= size
-    return data.broadcast(sizes={**sizes, **data.sizes}).copy()
+    ones = sc.DataArray(data=sc.ones(sizes=template.sizes))
```

Comment:
> Also we only want to handle pixel masks, not other masks.

Not sure I understand why pixel masks are special? Should we do the same if we have a mask in wavelength or anything else? Unless this is special handling for the SANS workflow, but we never intended to write a generic handling of masks in variance broadcast?

Reply:
What I mean is, we must only consider masks that relate to the dimensions being broadcasted. Other masks need to be excluded from the counting.

Reply:
> Should make ones only for the broadcast dims, otherwise this is wasteful.

I think this breaks down in the case where the mask is two-dimensional: one of the dims is also in the input, and the other is in the broadcast template?
I guess this may be a unusual case, so I should add special handling.